File: server/py/tlsprobe.py
"""Real TLS handshakes.

Verification is off throughout, and that is the point: an expired or
self-signed certificate is exactly what an assessment needs to see. Nothing
is ever written to these sockets.

The certificate comes back as DER and is decoded by the ASN.1 parser the
browser already has, so there is one parser rather than two that can
disagree.
"""

import base64
import socket
import ssl

VERSIONS = [
    ("TLSv1", ssl.TLSVersion.TLSv1),
    ("TLSv1.1", ssl.TLSVersion.TLSv1_1),
    ("TLSv1.2", ssl.TLSVersion.TLSv1_2),
    ("TLSv1.3", ssl.TLSVersion.TLSv1_3),
]


def _context(version=None):
    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE
    if version is not None:
        ctx.minimum_version = version
        ctx.maximum_version = version
        if version in (ssl.TLSVersion.TLSv1, ssl.TLSVersion.TLSv1_1):
            # OpenSSL 3 refuses legacy versions at the default security level.
            # Relaxed here solely so they can be tested *for*.
            try:
                ctx.set_ciphers("DEFAULT@SECLEVEL=0")
            except ssl.SSLError:
                pass
    return ctx


def handshake(host, port, servername=None, timeout=6.0, version=None):
    ctx = _context(version)
    sni = servername if servername and not servername.replace(".", "").isdigit() else None
    raw = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    raw.settimeout(timeout)
    try:
        raw.connect((host, port))
        with ctx.wrap_socket(raw, server_hostname=sni) as tls:
            der = tls.getpeercert(binary_form=True)
            cipher = tls.cipher() or (None, None, None)
            return {
                "ok": True,
                "protocol": tls.version(),
                "cipher": cipher[0],
                "cipherVersion": cipher[1],
                "alpn": tls.selected_alpn_protocol(),
                "certDer": base64.b64encode(der).decode("ascii") if der else None,
            }
    except ssl.SSLError as err:
        return {"ok": False, "reason": "ssl", "message": str(err)}
    except (socket.timeout, OSError) as err:
        return {"ok": False, "reason": "connect", "message": str(err)}
    finally:
        try:
            raw.close()
        except OSError:
            pass


def _local_refusal(result):
    """Errors raised by our own OpenSSL before anything reached the network."""
    text = ((result.get("reason") or "") + " " + (result.get("message") or "")).lower()
    return any(needle in text for needle in (
        "no protocols available", "unsupported protocol", "version too low",
        "no ciphers available", "wrong_version"))


def enumerate_protocols(host, port, servername=None, timeout=4.0):
    """Three outcomes per version, kept distinct.

    'untestable' means this Python build would not offer the version, which
    is a fact about the scanner, not the target. Reporting it as unsupported
    would be a result we never observed.
    """
    accepted, refused, untestable = [], [], []
    for label, version in VERSIONS:
        result = handshake(host, port, servername, timeout, version)
        if result["ok"]:
            accepted.append(label)
        elif _local_refusal(result):
            untestable.append(label)
        else:
            refused.append(label)
    return accepted, refused, untestable


def inspect(host, port, servername=None, timeout=6.0, enumerate_versions=True):
    primary = handshake(host, port, servername, timeout)
    if not primary["ok"]:
        return {"port": port, "reachable": False,
                "reason": primary["reason"], "message": primary.get("message")}

    if enumerate_versions:
        accepted, refused, untestable = enumerate_protocols(host, port, servername)
    else:
        accepted = [primary["protocol"]] if primary["protocol"] else []
        refused, untestable = [], [v[0] for v in VERSIONS]

    return {
        "port": port,
        "reachable": True,
        "negotiated": primary["protocol"],
        "protocols": accepted,
        "protocolsRefused": refused,
        "protocolsUntestable": untestable,
        "minProtocol": accepted[0] if accepted else primary["protocol"],
        "cipher": primary["cipher"],
        "alpn": primary["alpn"],
        "trusted": False,
        "trustError": "chain not validated — inspection mode",
        "certDer": primary["certDer"],
        "chainDer": [],
    }

File: server/lib/tls-probe.js
/* ============================================================================
   Real TLS handshakes.

   Certificate validation is switched off throughout this file, and that is
   the point: an expired or self-signed certificate is exactly what an
   assessment needs to see. No data is ever written to these sockets — they
   are opened, inspected and closed — so there is nothing to intercept.
   ========================================================================= */
'use strict';

const tls = require('tls');

const VERSIONS = ['TLSv1', 'TLSv1.1', 'TLSv1.2', 'TLSv1.3'];

/**
 * Complete one handshake and report what the peer presented.
 *
 * @param {string} host        address to connect to
 * @param {number} port
 * @param {object} [options]   servername, timeoutMs, version
 */
function handshake(host, port, options) {
  const opts = options || {};
  const timeoutMs = opts.timeoutMs || 6000;

  return new Promise((resolve) => {
    let settled = false;
    const finish = (value) => {
      if (settled) return;
      settled = true;
      try { socket.destroy(); } catch (e) { /* already gone */ }
      resolve(value);
    };

    const config = {
      host,
      port,
      rejectUnauthorized: false,
      timeout: timeoutMs,
      ALPNProtocols: ['h2', 'http/1.1']
    };
    // SNI only when we have a real name; sending an IP literal is invalid.
    if (opts.servername && !/^[\d.]+$/.test(opts.servername)) config.servername = opts.servername;
    if (opts.version) {
      config.minVersion = opts.version;
      config.maxVersion = opts.version;
      // OpenSSL 3 refuses legacy versions at the default security level. This
      // relaxation exists solely so the old versions can be *tested for*.
      if (opts.version === 'TLSv1' || opts.version === 'TLSv1.1') {
        config.ciphers = 'DEFAULT:@SECLEVEL=0';
      }
    }

    const socket = tls.connect(config, () => {
      const peer = socket.getPeerCertificate(true);
      const cipher = socket.getCipher() || {};
      finish({
        ok: true,
        protocol: socket.getProtocol(),
        cipher: cipher.name || null,
        cipherStandard: cipher.standardName || null,
        cipherVersion: cipher.version || null,
        alpn: socket.alpnProtocol || null,
        authorized: socket.authorized,
        authorizationError: socket.authorizationError ? String(socket.authorizationError) : null,
        certDer: derOf(peer),
        chainDer: chainOf(peer),
        chainLength: chainOf(peer).length + 1
      });
    });

    socket.once('timeout', () => finish({ ok: false, reason: 'timeout' }));
    socket.once('error', (err) => finish({
      ok: false,
      reason: (err && err.code) || 'error',
      message: err ? String(err.message) : null
    }));
  });
}

function derOf(cert) {
  return cert && cert.raw ? cert.raw.toString('base64') : null;
}

/** Intermediates the peer sent, stopping at the self-signed root. */
function chainOf(cert) {
  const out = [];
  let node = cert && cert.issuerCertificate;
  const seen = new Set();
  while (node && node.raw) {
    const key = node.fingerprint256 || node.raw.toString('base64').slice(0, 64);
    if (seen.has(key)) break;          // self-signed roots point at themselves
    seen.add(key);
    out.push(node.raw.toString('base64'));
    if (out.length >= 8) break;
    node = node.issuerCertificate;
  }
  return out;
}

/**
 * Determine which protocol versions the peer will negotiate.
 *
 * Three outcomes per version, kept distinct. "untestable" means this Node
 * build would not offer the version, which is a statement about the scanner
 * and not about the target — collapsing it into "unsupported" would be a
 * result we did not actually observe.
 */
async function enumerateProtocols(host, port, options) {
  const opts = options || {};
  const accepted = [];
  const refused = [];
  const untestable = [];

  for (const version of VERSIONS) {
    const result = await handshake(host, port, {
      servername: opts.servername,
      timeoutMs: opts.timeoutMs || 4000,
      version
    });
    if (result.ok) { accepted.push(version); continue; }
    if (isLocalRefusal(result)) untestable.push(version);
    else refused.push(version);
  }

  return { accepted, refused, untestable };
}

/* Errors raised by our own OpenSSL before anything reached the network. */
function isLocalRefusal(result) {
  const text = ((result.reason || '') + ' ' + (result.message || '')).toLowerCase();
  return text.indexOf('no protocols available') !== -1 ||
         text.indexOf('unsupported protocol') !== -1 ||
         text.indexOf('version too low') !== -1 ||
         text.indexOf('no ciphers available') !== -1 ||
         text.indexOf('err_ssl_version') !== -1;
}

/**
 * Full inspection of one TLS endpoint, shaped for CT's `asset.tls` field.
 * `cert` stays as DER; the client decodes it with CT.crypto.parseCertificate.
 */
async function inspect(host, port, options) {
  const opts = options || {};
  const primary = await handshake(host, port, opts);
  if (!primary.ok) return { port, reachable: false, reason: primary.reason, message: primary.message || null };

  const protocols = opts.enumerate === false
    ? { accepted: primary.protocol ? [primary.protocol] : [], refused: [], untestable: VERSIONS }
    : await enumerateProtocols(host, port, opts);

  return {
    port,
    reachable: true,
    negotiated: primary.protocol,
    protocols: protocols.accepted,
    protocolsRefused: protocols.refused,
    protocolsUntestable: protocols.untestable,
    minProtocol: protocols.accepted.length ? protocols.accepted[0] : primary.protocol,
    cipher: primary.cipherStandard || primary.cipher,
    alpn: primary.alpn,
    trusted: primary.authorized,
    trustError: primary.authorizationError,
    chainLength: primary.chainLength,
    certDer: primary.certDer,
    chainDer: primary.chainDer
  };
}

module.exports = { VERSIONS, handshake, enumerateProtocols, inspect };
